### ISSUE TITLE
TLS certificate verification is too lenient with wildcards

### ISSUE THREAD

**Issue by cespare** [dakrone/clj-http#298]
If you make an HTTPS request to a server that presents a certificate that's otherwise valid but doesn't match the host, clj-http will not give an exception.

I would expect that this should be an exception unless `:insecure? true` was provided.

I checked three TLS implementations and they all reject such certificates (I've changed the hostname in these examples):

**curl (OpenSSL)**

```
$ curl https://foo.bar.example.com
curl: (51) SSL: no alternative certificate subject name matches target host name 'foo.bar.example.com'
```

**Chrome**

    NET::ERR_CERT_COMMON_NAME_INVALID

**Go standard library**

    Get https://foo.bar.example.com: x509: certificate is valid for *.example.com, example.com, not foo.bar.example.com

**Comment by dakrone** [comment]
@cespare is this on 2.x or master/3.x branch?

**Comment by cespare** [comment]
I'm using 2.0.0, the latest release on Clojars. However, I just tried git master and it appears to give the same result.

**Comment by dakrone** [comment]
Hmm... that's very alarming, I'm looking at the code for it and I'm not doing
anything special, just using Apache's default
`SSLSocketFactory/getSocketFactory`.

@cespare can you tell me more about the request? I'm afraid I don't have a valid
certificate that I can grab for an invalid hostname (perhaps I can grab a valid
cert from a website and use it locally to test).

How are you issuing the request? Is it a standard get/put/post/whatever? Are you
using a proxy or any special connection manager settings?

**Comment by dakrone** [comment]
Also, looking at the implementation of Apache's `AbstractVerifier.verify(...)`, it does compare hostnames for certificates

**Comment by cespare** [comment]
OK, I've confirmed it does do some checking; the problem is specifically with wildcards.

You can test using some valid HTTPS website; I'll use Google (ip 173.194.121.34). If you just try connecting use a bad name, say example.com:

```
$ echo '173.194.121.34 example.com' | sudo tee -a /etc/hosts                                                      
173.194.121.34 example.com
$ curl https://example.com
curl: (51) SSL: no alternative certificate subject name matches target host name 'example.com'
```

clj-http will also throw an exception:

```
user=> (require '[clj-http.client :as http])
nil
user=> (http/get "https://example.com")
SSLPeerUnverifiedException Host name 'example.com' does not match the certificate subject provided by the peer (CN=*.google.com, O=Google Inc, L=Mountain View, ST=California, C=US)  org.apache.http.conn.ssl.SSLConnectionSocketFactory.verifyHostname (SSLConnectionSocketFactory.java:465)
```

If you add a two-level subdomain of google.com, curl also won't accept that:

```
$ echo '173.194.121.34 foo.bar.google.com' | sudo tee -a /etc/hosts
173.194.121.34 foo.bar.google.com
$ curl https://foo.bar.google.com
curl: (51) SSL: no alternative certificate subject name matches target host name 'foo.bar.google.com'
```

That's because the Google certificate is valid on `*.google.com`, but not `*.bar.google.com`. However, clj-http will allow it:

```
(http/get "https://foo.bar.google.com")
ExceptionInfo clj-http: status 404  clj-http.client/wrap-exceptions/fn--4810 (client.clj:196)
```

**Comment by cespare** [comment]
I've updated the issue title accordingly.

**Comment by cespare** [comment]
@dakrone I sort of figured from the beginning that this would be an apache bug, but was dreading whatever jira process I'd have to go through to file a ticket. If you confirm this is on their end I can try to write a Java repro and take it up with them.

**Comment by dakrone** [comment]
@cespare I have reproduced this with the following program:

``` Java
import java.util.Arrays;
import java.util.Locale;

// Most of the methods for this are copied directly from httpclient 4.5 AbstractVerifier.java
public class HostnameMatcher {

    final static String[] BAD_COUNTRY_2LDS =
            { "ac", "co", "com", "ed", "edu", "go", "gouv", "gov", "info",
                    "lg", "ne", "net", "or", "org" };

    static {
        // Just in case developer forgot to manually sort the array.  :-)
        Arrays.sort(BAD_COUNTRY_2LDS);
    }

    /**
     * Counts the number of dots "." in a string.
     * @param s  string to count dots from
     * @return  number of dots
     */
    public static int countDots(final String s) {
        int count = 0;
        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) == '.') {
                count++;
            }
        }
        return count;
    }

    private static boolean validCountryWildcard(final String parts[]) {
        if (parts.length != 3 || parts[2].length() != 2) {
            return true; // it's not an attempt to wildcard a 2TLD within a country code
        }
        return Arrays.binarySearch(BAD_COUNTRY_2LDS, parts[1]) < 0;
    }

    private static boolean matchIdentity(final String host, final String identity, final boolean strict) {
        if (host == null) {
            return false;
        }
        final String normalizedHost = host.toLowerCase(Locale.ROOT);
        final String normalizedIdentity = identity.toLowerCase(Locale.ROOT);
        // The CN better have at least two dots if it wants wildcard
        // action.  It also can't be [*.co.uk] or [*.co.jp] or
        // [*.org.uk], etc...
        final String parts[] = normalizedIdentity.split("\\.");
        final boolean doWildcard = parts.length >= 3 && parts[0].endsWith("*") &&
            (!strict || validCountryWildcard(parts));
        if (doWildcard) {
            boolean match;
            final String firstpart = parts[0];
            if (firstpart.length() > 1) { // e.g. server*
                final String prefix = firstpart.substring(0, firstpart.length() - 1); // e.g. server
                final String suffix = normalizedIdentity.substring(firstpart.length()); // skip wildcard part from cn
                final String hostSuffix = normalizedHost.substring(prefix.length()); // skip wildcard part from normalizedHost
                match = normalizedHost.startsWith(prefix) && hostSuffix.endsWith(suffix);
            } else {
                match = normalizedHost.endsWith(normalizedIdentity.substring(1));
            }
            return match && (!strict || countDots(normalizedHost) == countDots(normalizedIdentity));
        } else {
            return normalizedHost.equals(normalizedIdentity);
        }
    }


    public static void main(String[] args) {
        if (args.length != 3) {
            System.out.println("Usage: java HostnameMatcher <host> <cert-wildcard> <strict>");
            System.exit(1);
        }
        String host = args[0];
        String cert = args[1];
        boolean strict = "true".equals(args[2]) ? true : false;
        System.out.println("Host: " + host);
        System.out.println("Cert: " + cert);
        System.out.println("Strict? " + strict);
        boolean match = matchIdentity(host, cert, strict);
        System.out.println("--> Match? [" + match + "]");
    }
}
```

And then running it:

```
~/src/java λ javac HostnameMatcher.java
~/src/java λ java HostnameMatcher foo.bar.google.com "*.google.com" false
Host: foo.bar.google.com
Cert: *.google.com
Strict? false
--> Match? [true]
~/src/java λ java HostnameMatcher foo.bar.google.com "*.google.com" true
Host: foo.bar.google.com
Cert: *.google.com
Strict? true
--> Match? [false]
```

It looks like we need to do whatever is necessary to turn on the strict hostname
matching in the Apache http client library. I'll look into doing this.

Thanks for bringing this to my attention!

**Comment by dakrone** [comment]
@cespare I have pushed a commit to the 2.x branch that fixes this for me, can you confirm that this fixes the problem for you? Once you do I will push a new release of clj-http (I will probably also add the option to revert to the old behavior if desired)

**Comment by dakrone** [comment]
(I also think it is insane that this is not the default in the Apache HTTP library...)

**Comment by cespare** [comment]
@dakrone confirmed! I presume you'll also apply this patch to master?

Thanks for the quick turnaround!

**Comment by dakrone** [comment]
@cespare I have pushed a commit to master as well to fix this, and I have released clj-http `2.0.1` and another `3.0.0-SNAPSHOT` with this fixed. Thanks again for bringing this to my attention!

**Comment by cespare** [comment]
@dakrone Thanks for fixing this so quickly.
